tests/test_enums.py

```diff
@@ -0,0 +1,275 @@
+from typing import Any
+from unittest import mock
+
+import pytest
+
+from disnake import enums
+
+
+@mock.patch.object(enums.EnumMeta, "__is_enum_instantiated__", new=False)
+def test_init_first_enum():
+    assert enums.EnumMeta.__is_enum_instantiated__ is False
+
+    class Enum(metaclass=enums.EnumMeta):
+        pass
+
+    assert enums.EnumMeta.__is_enum_instantiated__ is True
+
+
+def test_init_followup_enum():
+    assert enums.EnumMeta.__is_enum_instantiated__ is True
+
+    class TestEnum(int, enums.Enum):
+        pass
+
+    assert enums.EnumMeta.__is_enum_instantiated__ is True
+
+    assert TestEnum.__mro__ == (TestEnum, int, enums.Enum, object)
+
+
+def test_init_enum_incorrect_base_number():
+    with pytest.raises(TypeError, match="at most two base classes"):
+
+        class FailedEnum(int, str, enums.Enum):
+            pass
+
+
+def test_init_enum_incorrect_base_order():
+    with pytest.raises(TypeError, match="first base class"):
+
+        class FailedEnum(enums.Enum, int):
+            pass
+
+
+def test_init_enum_untyped_no_members():
+    class UntypedEnum(enums.Enum):
+        _private = "a"
+
+    assert UntypedEnum._private == "a"
+
+
+def test_init_enum_untyped_inheritance():
+    with pytest.raises(TypeError):
+
+        class UntypedEnum(enums.Enum):
+            member = "a"
+
+
+def test_init_enum_inheritance():
+    class UntypedBaseEnum(enums.Enum):
+        def x(self):
+            return "woo"
+
+    class TypedInheritor(int, UntypedBaseEnum):
+        y = 1
+
+    assert TypedInheritor.__base_type__ is int  # type: ignore
+    assert TypedInheritor.y.x() == "woo"
+
+
+def test_init_enum_typed_inheritance():
+    class TypedBaseEnum(int, enums.Enum):
+        def x(self):
+            return "woo"
+
+    class UntypedInheritor(TypedBaseEnum):
+        # Base type <int> inferred from TypedBaseEnum
+        y = 1
+
+    assert UntypedInheritor.__base_type__ is int  # type: ignore
+    assert UntypedInheritor.y.x() == "woo"
+
+
+def test_init_enum_member_inheritance():
+    class BaseEnum(int, enums.Enum):
+        a = 1
+
+    with pytest.raises(TypeError):
+
+        class ChildEnum(BaseEnum):  # type: ignore
+            b = 2
+
+
+@pytest.mark.parametrize("name", ["mro", "name", "value"])
+def test_init_enum_illegal_name(name: str):
+    with pytest.raises(ValueError, match="Invalid Enum member name"):
+        # TODO: come up with a better way of running this test.
+        exec(f"class PainEnum(int, enums.Enum):\n\t{name} = 1\n", {"enums": enums})  # noqa: S102
```

Comment:
```suggestion
    def init_namespace(ns: Dict[str, Any]) -> None:
        ns[name] = 1

    with pytest.raises(ValueError, match="Invalid Enum member name"):
        types.new_class("PainEnum", (int, enums.Enum), {}, init_namespace)
```
Came up with a better way, after staring at the docs for a bit :smile:
The fourth parameter to [`types.new_class`](https://docs.python.org/3/library/types.html#types.new_class) essentially replaces the class body, so this'll end up raising on the `ns[name] = 1` assignment as it should.
Using `types.prepare_class` would also work:
```py
meta, ns, kwds = types.prepare_class("PainEnum", (int, enums.Enum), {})
ns[name] = 1
```
But might as well do it properly and (try to) instantiate the full class :p

Reply:
That's much better, thanks ^^
Done as part of 726c82657a838535ce0e71a3e0bbf6bd5fec5141.